### PR TITLE
Abstracted ConcordanceWalker

### PR DESCRIPTION
**Pull request by davidbenjamin**
@takutosato Much of this code will be familiar because you wrote it.  Most of the files changed are just moving classes to the validations package.

I couldn't see a good way to have a sufficiently flexible tool without outputting variants as a vcf and then relying on `VariantsToTable`.  In particular, the CRSP validation relies on non-standard annotations for the pooled bam depth and the expected allele fraction.  Also, if we use this tool to make training data for machine learning classifiers we'll want to keep as many covariates as we can.  I therefore replaced `VariantStatus___` with output to vcf, but kept the commit separate in case you see another option.

I wrote the CRSP wdl that uses all this on a separate branch and am currently testing it.  Once that works I'll issue its PR.

### CODE REVIEW

**Comment by takutosato** [comment]
Yea, I printing out a vcf as an output makes much more sense.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/validation/ConcordanceIntegrationTest.java [hunk 1]
```diff
@@ -65,15 +58,16 @@ public void testt22X() throws Exception {
         final String testDir = publicTestDir + "org/broadinstitute/hellbender/tools/concordance/";
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/validation/ConcordanceIntegrationTest.java]: This is my typo, but there is an extra t in `testt22X()`

File: src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+                return TruthVersusEval.falseNegative(truthIterator.next());
+            } else if (areVariantsAtSameLocusConcordant(truthIterator.peek(), evalIterator.peek())) {
+                return TruthVersusEval.truePositive(truthIterator.next(), evalIterator.next());
+            } else {
+                // advance truth in case of same-locus discordance -- we could equally well advance eval
+                return TruthVersusEval.falseNegative(truthIterator.next());
+            }
+        }
+
+        public Spliterator<TruthVersusEval> spliterator() {
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java]: This method should be `private`.

File: src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    public static final String TRUTH_VARIANTS_SHORT_NAME = "truth";
+    public static final String EVAL_VARIANTS_SHORT_NAME = "eval";
+
+    public static final String CONFIDENCE_REGION_LONG_NAME = "confidence";
+    public static final String CONFIDENCE_REGION_SHORT_NAME = "C";
+
+    // The distance in bases to look ahead and cache when querying feature sources.
+    public static final int CACHE_LOOKAHEAD = 100_000;
+
+    @Argument(shortName = TRUTH_VARIANTS_SHORT_NAME, doc = "A VCF containing truth variants", optional = false)
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java]: done

File: src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            throw new GATKException("Header for " + source.getName() + " is not in VCF header format");
+        }
+        return (VCFHeader)header;
+    }
+
+    /**
+     * encapsulate the iteration over truth and eval.  hasNext() returns true if either the truth or eval iterator have
+     * any variants remaining.  next() produces variants in genomic order, along with their concordance state of
+     * false positive for an eval variant without concordant truth variant, false negative for a truth variant without
+     * a concordant eval variant, and true positive for concordance eval and truth variants.
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+     *
+     * This implementation correctly consumes a single next() of the wrapped iterators for false positives and
+     * false negatives and consumes a next() for both iterators in case of a true positive.
+     */
+    private class ConcordanceIterator implements Iterator<TruthVersusEval> {
+        private final PeekableIterator<VariantContext> truthIterator;
+        private final PeekableIterator<VariantContext> evalIterator;
+
+        public ConcordanceIterator(final  Iterator<VariantContext> truthIterator, final Iterator<VariantContext> evalIterator) {
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java]: done

File: src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+            this.truthIterator = new PeekableIterator<VariantContext>(truthIterator);
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java]: done

File: src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java (hunk 2 above)
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/engine/AbstractConcordanceWalker.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/Concordance.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    protected File truePositivesAndFalseNegativesVcf = null;
+
+    @Argument(doc = "A vcf to write true positives and false positives",
+            fullName = TRUE_POSITIVES_AND_FALSE_POSITIVES_LONG_NAME,
+            shortName = TRUE_POSITIVES_AND_FALSE_POSITIVES_SHORT_NAME,
+            optional = true)
+    protected File truePositivesAndFalsePositivesVcf = null;
+
+    // we count true positives, false positives, false negatives for snps and indels
+    private final EnumMap<ConcordanceState, MutableLong> snpCounts = new EnumMap<ConcordanceState, MutableLong>(ConcordanceState.class);
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/validation/Concordance.java]: done

**Pull request merged**